src/outpack/location_pull.py

```diff
@@ -0,0 +1,463 @@
+            )
+            _location_pull_files_archive(packet.packet, store, root)
+
+        mark_known(
+            root, packet.packet, LOCATION_LOCAL, packet.hash, time.time()
+        )
+
+    print(
+        f"Unpacked {n_packets} {pl(n_packets, 'packet')} in "
+        f"{humanize.time.precisedelta(int(time.time() - time_start))}."
+    )
+
+    cleanup()
+    return list(plan.packets.keys())
+
+
+# This approach may be suboptimal in the case where the user does not
+# already have a file store, as it means that files will be copied
+# around and hashed more than ideal:
+#
+# * hash the candidate file
+# * rehash on entry into the file store
+# * copy into the file store
+# * copy from the file store into the final location
+#
+# So in the case where a hash is only present once in a chain of
+# packets being pulled this will be one too many hashes and one too
+# many copies.
+#
+# However, this approach makes the logic fairly easy to deal with,
+# and copes well with data races and corruption of data on disk
+# (e.g., users having edited files that we rely on, or editing them
+# after we hash them the first time).
+def _location_pull_files(
+    files: List[PacketFileWithLocation], root: OutpackRoot
+) -> Tuple[FileStore, Callable[[], None]]:
+    store = root.files
+    if store is not None:
+
+        def cleanup():
+            return None
+
+        exists, missing = partition(lambda file: store.exists(file.hash), files)
+
+        if len(exists) > 0:
+            print(
+                f"Found {len(exists)} {pl(exists, 'file')} in the "
+                f"file store"
+            )
+    else:
+        print("Looking for suitable files already on disk")
+        store = _temporary_filestore(root)
+
+        def cleanup():
+            store.destroy()
+
+        for file in files:
+            path = find_file_by_hash(root, file.hash)
+            if path is not None:
+                store.put(path, file.hash)
+
+    if len(files) == 0:
+        print("All files available locally, no need to fetch any")
+    else:
+        locations = {file.location for file in files}
+        total_size = humanize.naturalsize(sum(file.size for file in files))
+        print(
+            f"Need to fetch {len(files)} {pl(files, 'file')} "
+            f"({total_size}) from {len(locations)} "
+            f"{pl(locations, 'location')}"
+        )
+        for location in locations:
+            from_this_location = [
+                file for file in files if file.location == location
+            ]
+            _location_pull_hash_store(
+                from_this_location,
+                location,
+                _location_driver(location, root),
+                store,
+            )
+
+    return store, cleanup
+
+
+def _location_pull_hash_store(
+    files: List[PacketFileWithLocation],
+    location_name: str,
+    driver,
+    store: FileStore,
+):
+    no_of_files = len(files)
+    # TODO: show a nice progress bar for users
```

Comment:
yeah - in general it would be nice to find a suitable python `cli`-like packages for all the print statements and things like progress bars

Reply:
Added a link to this into [mrc-4968](https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4968)